library/core/src/ffi/c_str.rs

```diff
@@ -780,8 +780,15 @@ const unsafe fn const_strlen(ptr: *const c_char) -> usize {
 pub struct Bytes<'a> {
     // since we know the string is nul-terminated, we only need one pointer
     ptr: NonNull<u8>,
-    phantom: PhantomData<&'a u8>,
+    phantom: PhantomData<&'a [c_char]>,
```

Comment:
What does this change add?

Reply:
It doesn't have any practical consequence. The new code has exactly the same variance and autotraits (at least built-in autotraits) as the old one. But different people prefer to use `PhantomData<T>` differently:

1. Some people set `T` to the _shortest_ possibility out of the set of all types that have the required variance and autotraits properties, or reach for a small set of ready-made idioms (usually short, which is why I am lumping this together). For example it's common to see `PhantomData<fn(U)>` (contravariant) or `PhantomData<fn(U) -> U>` (invariant) in data structures which have absolutely nothing to do with function pointers. These people decide what properties they want, then pick a type which hopefully delivers those properties.

2. Other people do the reverse: pick a type which mirrors what their data structure does the most closely, and hope that the correct variance and autotraits emerge from that for their data structure as a consequence. For example you might see `PhantomData<MutexGuard<'a, T>>` in a binding for some foreign language's mutex guard type, and the author can't even necessarily name what the resulting variances are.

3. A rare subgroup of both of the above is people who actually test the variance they're getting.

    ```rust
    fn f<'short, 'long: 'short>(x: MyType<'short>) -> MyType<'long> {
        x
    }

    fn g<'short, 'long: 'short>(x: MyType<'long>) -> MyType<'short> {
        x //~ lifetime may not live long enough
    }
    ```

4. The largest group: don't know / don't consider anything about variance and write whatever makes compilation succeed, even if it has incorrect variance.

`PhantomData<&'a u8>` is indicative of category 1 (_or_ 4, though I hope this would be unusual in standard library code) while `PhantomData<&'a [c_char]>` resembles category 2. I don't have an opinion about whether either is a better approach. 1 and 2 just have different failure modes.

This would make a bigger difference if support for nonstandard auto traits (https://github.com/rust-lang/rust/issues/13231) is ever stabilized. At that point all PhantomData usage would need to be audited, along the lines of https://github.com/rust-lang/rust/issues/13231#issuecomment-1399386472.